File: src/main/java/com/machfour/macros/core/MacrosEntity.java
package com.machfour.macros.core;

import org.jetbrains.annotations.NotNull;

import java.util.Map;

import static java.lang.Double.NaN;

/**
 * Defines common methods for each object to be persisted
 */

public interface MacrosEntity<M extends MacrosEntity<M>> {

    long NO_ID = -100;
    // special ID for the 'null' serving of just grams / mL
    long UNIT_SERVING = -101;
    long NO_DATE = -99;
    double UNSET = NaN;

    @NotNull
    Long getId();

    default boolean hasId() {
        return getId() != NO_ID;
    }

    @NotNull
    Long createTime();

    @NotNull
    Long modifyTime();

    @NotNull
    ObjectSource getObjectSource();

    // Used to get data by column objects
    <J> J getData(Column<M, J> c);

    boolean hasData(Column<M, ?> c);

    // Creates a mapping of column objects to their values for this instance
    ColumnData<M> getAllData(boolean readOnly);
    // equivalent to getAllData(true)
    ColumnData<M> getAllData();

    Table<M> getTable();
    Factory<M> getFactory();

    // ... Alternative methods that can be used with unique columns
    <N extends MacrosEntity<N>, J> void setFkParentNaturalKey(Column.Fk<M, ?, N> fkCol, Column<N, J> parentNaturalKey, N parent);
    <N, J> void setFkParentNaturalKey(Column.Fk<M, ?, N> fkCol, Column<N, J> parentNaturalKey, J data);
    <N> ColumnData<N> getFkParentNaturalKey(Column.Fk<M, ?, N> fkCol);
    Map<Column.Fk<M, ?, ?>, ?> getFkNaturalKeyMap();
    void copyFkNaturalKeyMap(MacrosEntity<M> from);
}

File: src/main/java/com/machfour/macros/core/MacrosEntity.kt
package com.machfour.macros.core

import com.machfour.macros.orm.Factory
import com.machfour.macros.orm.ObjectSource
import com.machfour.macros.sql.Column
import com.machfour.macros.sql.RowData
import com.machfour.macros.sql.Table

/**
 * Defines common methods for each object to be persisted
 */
interface MacrosEntity<M : MacrosEntity<M>> {
    companion object {
        const val NO_ID: Long = -100

        // special ID for the 'null' serving of just grams / mL
        const val UNIT_SERVING: Long = -101
        const val NO_DATE: Long = -99
        const val UNSET = Double.NaN

        fun <M: MacrosEntity<M>> Factory<M>.cloneWithoutMetadata(obj: MacrosEntity<M>): M {
            return construct(obj.dataCopy(withMetadata = false), ObjectSource.COMPUTED)
        }
    }

    val id: Long
    val hasId: Boolean
        get() = (id != NO_ID)

    val createTime: Long
    val modifyTime: Long
    val source: ObjectSource

    // Used to get data by column objects
    fun <J> getData(col: Column<M, J>): J?
    fun hasData(col: Column<M, *>): Boolean

    val data: RowData<M>
    val table: Table<M>
    val factory: Factory<M>

    fun dataCopy(withMetadata: Boolean): RowData<M>
    fun dataFullCopy(): RowData<M> = dataCopy(withMetadata = true)

    // ... Alternative methods that can be used with unique columns
    fun <N : MacrosEntity<N>, J> setFkParentNaturalKey(fkCol: Column.Fk<M, *, N>, parentNaturalKey: Column<N, J>, parent: N)
    fun <N, J> setFkParentNaturalKey(fkCol: Column.Fk<M, *, N>, parentNaturalKey: Column<N, J>, data: J)
    fun <N> getFkParentNaturalKey(fkCol: Column.Fk<M, *, N>): RowData<N>
    val fkNaturalKeyMap: Map<Column.Fk<M, *, *>, *>
    fun copyFkNaturalKeyMap(from: MacrosEntity<M>)

}